crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4804,6 +4804,41 @@ impl<'db> TypeInferenceBuilder<'db> {
     ) -> Type<'db> {
         let parameters = &*subscript.slice;
         match known_instance {
+            KnownInstanceType::Annotated => {
+                let mut report_invalid_parameters = || {
+                    self.diagnostics.add_lint(
+                        &INVALID_TYPE_PARAMETER,
+                        subscript.into(),
+                        format_args!(
+                            "Type `{}` expected at least 2 parameters (one type and one annotation)",
+                            known_instance.repr(self.db)
+                        ),
+                    );
+                };
+
+                let ast::Expr::Tuple(ast::ExprTuple { elts, .. }) = parameters else {
+                    report_invalid_parameters();
+
+                    // `Annotated[]` with less than two parameters is an error at runtime.
+                    // However, we still treat `Annotated[T]` as `T` here for the purpose of
+                    // giving better diagnostics later on.
+                    // Pyright also does this. Mypy doesn't and fallback to `Any` instead.
+                    return self.infer_type_expression(parameters);
+                };
+
+                if elts.len() < 2 {
+                    report_invalid_parameters();
+                }
+
+                let [first, ..] = &elts[..] else {
+                    self.infer_type_expression(parameters);
+                    return Type::Unknown;
+                };
+
+                let ty = self.infer_type_expression(first);
+                self.store_expression_type(parameters, ty);
```

Comment:
We currently have a requirement that we store a type for every expression (in order to support hover in an LSP). I suspect the problem you're running into is that you never infer or store a type for the second or subsequent parameters. I think you will need to infer the first as a type expression, infer the rest of the parameters as value expressions, and then return the type of the first and ignore the rest.

Reply:
(FYI, this is the same thing that caused a panic in the first version of https://github.com/astral-sh/ruff/pull/14858/, @InSyncWithFoo)